src/spartan/upsnark.rs

```diff
@@ -0,0 +1,482 @@
+
+      assert_eq!(evals_A.len(), evals_B.len());
+      assert_eq!(evals_A.len(), evals_C.len());
+
+      let span_e = tracing::span!(tracing::Level::TRACE, "eval_combo_old");
+      let _enter_e = span_e.enter();
+      let r_sq = r * r;
+      let thing = (0..evals_A.len())
+        .into_par_iter()
+        .map(|i| {
+          evals_A[i] + evals_B[i] * r + evals_C[i] * r_sq
+        })
+        .collect::<Vec<G::Scalar>>();
+      drop(_enter_e);
+      drop(span_e);
+
+      thing
+    };
+    drop(_enter);
+    drop(span);
+
+    let poly_z = {
+      z.resize(pk.S.num_vars * 2, G::Scalar::ZERO);
+      z
+    };
+
+    let comb_func = |poly_A_comp: &G::Scalar, poly_B_comp: &G::Scalar| -> G::Scalar {
+      *poly_A_comp * *poly_B_comp
+    };
+    let (sc_proof_inner, r_y, _claims_inner) = SumcheckProof::prove_quad(
+      &claim_inner_joint,
+      num_rounds_y,
+      &mut MultilinearPolynomial::new(poly_ABC),
+      &mut MultilinearPolynomial::new(poly_z),
+      comb_func,
+      &mut transcript,
+    )?;
+
+    let span = tracing::span!(tracing::Level::TRACE, "MultilinearPolynomial::evaluate_with");
+    let _enter = span.enter();
+    let eval_W = MultilinearPolynomial::evaluate_with(&W.W, &r_y[1..]);
+    drop(_enter);
+    drop(span);
+
+    let eval_arg = EE::prove(
+      &pk.ck,
+      &pk.pk_ee,
+      &mut transcript,
+      &u.comm_W,
+      &W.W.clone(),
+      &r_y[1..].to_vec(),
+      &eval_W,
+    )?;
+
+    Ok(R1CSSNARK {
+      comm_W: u.comm_W.compress(),
+      sc_proof_outer,
+      claims_outer: (claim_Az, claim_Bz, claim_Cz),
+      sc_proof_inner,
+      eval_W,
+      eval_arg,
+    })
+  }
+
+  /// verifies a proof of satisfiability of a `RelaxedR1CS` instance
+  #[tracing::instrument(skip_all, name = "SNARK::verify")]
+  fn verify(&self, vk: &Self::VerifierKey, io: &[G::Scalar]) -> Result<(), SpartanError> {
+    // construct an instance using the provided commitment to the witness and IO
+    let comm_W = Commitment::<G>::decompress(&self.comm_W)?;
+    let u = R1CSInstance::new(&vk.S, &comm_W, io)?;
+
+    let mut transcript = G::TE::new(b"R1CSSNARK");
+
+    // append the digest of R1CS matrices and the RelaxedR1CSInstance to the transcript
+    transcript.absorb(b"vk", &vk.digest());
+    transcript.absorb(b"U", &u);
+
+    let (num_rounds_x, num_rounds_y) = (
+      usize::try_from(vk.S.num_cons.ilog2()).unwrap(),
+      (usize::try_from(vk.S.num_vars.ilog2()).unwrap() + 1),
+    );
+
+    // outer sum-check
+    let tau = (0..num_rounds_x)
+      .map(|_i| transcript.squeeze(b"t"))
+      .collect::<Result<Vec<G::Scalar>, SpartanError>>()?;
+
+    let (claim_outer_final, r_x) =
+      self
+        .sc_proof_outer
+        .verify(G::Scalar::ZERO, num_rounds_x, 3, &mut transcript)?;
+
+    // verify claim_outer_final
+    let (claim_Az, claim_Bz, claim_Cz) = self.claims_outer;
+    let taus_bound_rx = EqPolynomial::new(tau).evaluate(&r_x);
```

Comment:
Add the `evaluate_with_chi` function to avoid recomputing lagrange basis for r_x.